src/JustEat.StatsD/SpanStatsDMessageFormatter.cs

```diff
@@ -0,0 +1,266 @@
+        public static ref FixedBuffer Add(ref this FixedBuffer src, byte[] str)
+        {
+            str.CopyTo(src.Buffer.Slice(src.Position));
+            src.Position += str.Length;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, byte str)
+        {
+            src.Buffer[src.Position] = str;
+            src.Position += 1;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, string str)
+        {
+            var written = Encoding.UTF8.GetBytes(str, src.Buffer.Slice(src.Position));
+            src.Position += written;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, double value)
+        {
+            Span<char> dest = stackalloc char[32];
+            value.TryFormat(dest, out int written, "", CultureInfo.InvariantCulture);
+
+            
+
+            ref var pos = ref src.Position;
+            
+            for (var i = 0; i < written; i++)
+            {
+                src.Buffer[pos++] = (byte) dest[i];
+            }
+
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, long value)
+        {
+            if (!Utf8Formatter.TryFormat(value, src.Buffer.Slice(src.Position), out int written, new StandardFormat('d')))
+            {
+                throw new OutOfMemoryException();
+            }
+
+            src.Position += written;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Clear(ref this FixedBuffer src)
+        {
+            src.Position = 0;
+        }
+    }
+
+    public ref struct FixedBuffer
+    {
+        public Span<byte> Buffer;
+        public int Position;
+
+        public FixedBuffer(Span<byte> buffer) : this()
+        {
+            Buffer = buffer;
+            Position = 0;
+        }
+    }
+
+    public class SpanStatsDMessageFormatter
+    {
+        private const double DefaultSampleRate = 1.0;
+
+        [ThreadStatic]
+        private static Random _random;
+
+        private readonly byte[] _prefix;
+
+        public SpanStatsDMessageFormatter()
+            : this(string.Empty) {}
+
+        public SpanStatsDMessageFormatter(string prefix)
+        {
+            _prefix = !string.IsNullOrWhiteSpace(prefix) ?
+                Encoding.UTF8.GetBytes(prefix + ".") :
+                Array.Empty<byte>();
+        }
+
+        private static Random Random => _random ?? (_random = new Random());
+
+        public void Timing(long milliseconds, string statBucket, ref FixedBuffer fixedBuffer)
```

Comment:
I'm not sure we'd want to create a new public type with all these "advanced" overloads using a custom struct.